lib/schedule/data.ex

```diff
@@ -66,34 +80,75 @@ defmodule Schedule.Data do
           hastus: %{String.t() => binary()}
         }
 
-  @spec all_routes(t()) :: [Route.t()]
-  def all_routes(%__MODULE__{routes: routes}), do: routes
+  @table_schema [
+    {:routes, [:id, :route], []},
+    {:route_patterns, [:id, :route_pattern], []},
+    {:timepoints_by_route, [:route_id, :timepoints], []},
+    {:timepoint_names_by_id, [:id, :timepoint_name], []},
+    {:shapes, [:route_id, :shapes], []},
+    {:stops, [:id, :stop], []},
+    {:trips, [:id, :service_id, :trip], [:service_id]},
+    {:blocks, [:id, :service_id, :block], [:service_id]},
+    {:calendar, [:date, :service_ids], []},
+    {:runs, [:id, :service_id, :run], [:service_id]},
+    {:swings, [:service_id_and_route_id, :swings], []}
+  ]
+
+  @spec all_routes(tables()) :: [Route.t()]
+  def all_routes(%{routes: routes_table}) do
+    :mnesia.ets(fn ->
+      routes_table
+      |> :ets.tab2list()
+      |> Enum.map(fn {_route_id, route} -> route end)
+    end)
+  end
 
-  @spec timepoints_on_route(t(), Route.id()) :: [Timepoint.t()]
-  def timepoints_on_route(
-        %__MODULE__{timepoints_by_route: timepoints_by_route},
-        route_id
-      ),
-      do: Map.get(timepoints_by_route, route_id, [])
+  @spec timepoints_on_route(tables(), Route.id()) :: [Timepoint.t()]
+  def timepoints_on_route(%{timepoints_on_route: timepoints_on_route_table}, route_id) do
+    case :mnesia.dirty_read(timepoints_on_route_table, route_id) do
+      [{^timepoints_on_route_table, _, timepoints}] -> timepoints
+      _ -> []
+    end
+  end
 
-  @spec timepoint_names_by_id(t()) :: Timepoint.timepoint_names_by_id()
-  def timepoint_names_by_id(%__MODULE__{timepoint_names_by_id: timepoint_names_by_id}),
-    do: timepoint_names_by_id
+  @spec timepoint_names_by_id(tables()) :: Timepoint.timepoint_names_by_id()
+  def timepoint_names_by_id(%{timepoint_names_by_id: timepoint_names_by_id_table}) do
+    :mnesia.ets(fn ->
+      timepoint_names_by_id_table
+      |> :ets.tab2list()
+      |> Map.new()
+    end)
+  end
 
-  @spec stop(t(), Stop.id()) :: Stop.t() | nil
-  def stop(%__MODULE__{stops: stops}, stop_id), do: stops[stop_id]
+  @spec stop(tables(), Stop.id()) :: Stop.t() | nil
+  def stop(%{stops: stops_table}, stop_id) do
+    case :mnesia.dirty_read(stops_table, stop_id) do
+      [{^stops_table, _, stop}] -> stop
+      _ -> nil
+    end
+  end
 
-  @spec trip(t(), Schedule.Trip.id()) :: Schedule.Trip.t() | nil
-  def trip(%__MODULE__{trips: trips}, trip_id), do: trips[trip_id]
+  @spec trip(tables(), Schedule.Trip.id()) :: Schedule.Trip.t() | nil
+  def trip(%{trips: trips_table}, trip_id) do
+    case :mnesia.dirty_read(trips_table, trip_id) do
+      [{^trips_table, _, trip}] -> trip
+      _ -> nil
+    end
+  end
 
-  @spec trips_by_id(t(), [Schedule.Trip.id()]) :: %{Schedule.Trip.id() => Schedule.Trip.t()}
-  def trips_by_id(%__MODULE__{trips: trips}, trip_ids) do
-    Map.take(trips, trip_ids)
+  # TODO: can this be done in a more performant way?
```

Comment:
Yes, you can pass a list of selectors to `:mnesia.dirty_select`.

Reply:
Hmm, I didn't see an obvious way of doing that from reading about [match specifications](https://www.erlang.org/doc/apps/erts/match_spec.html) since there doesn't seem to be a "value is in list" operator, but I suppose it's possible to just string together a bunch of equality checks with `or` statements.

Reply:
`select` and `dirty_select` take a list of selectors (https://erlang.org/doc/man/mnesia.html#select-2), so you can do something like:
```elixir
selectors = for trip_id <- trips_ids do
  ...generate selector...
end
results = :mnesia.dirty_select(trips, selectors)
```